File: Diablo_2_Web_Lobby/ladder/LadderParser.py
from struct import unpack,calcsize
from os import stat


### Some struct from d2cs/d2dbs source
#
# ladder header (4 + 4 = 8):
#   bn_int maxtype
#   bn_int checksum
LD_HEAD="<2i"
szLD_HEAD = calcsize(LD_HEAD)

#
# ladder info (4 + 2 + 1 + 1 + 16 = 24):
#   bn_int experience
#   bn_short status
#   bn_byte level
#   bn_byte class;
#   char charname[16];
LD_INFO="<Ihbb16s"
szLD_INFO = calcsize(LD_INFO)

#
# ladder index (4 + 4 + 4 = 12):
#   bn_int type
#   bn_int offset
#   bn_int number
LD_INDEX="<3i"
szLD_INDEX = calcsize(LD_INDEX)

## Status flags
S_INIT = 0x1
S_EXP  = 0x20
S_HC   = 0x04
S_DEAD = 0x08


classes = {
    0x00 : ['Amazon', 'f'],
    0x01 : ['Sorceress', 'f'],
    0x02 : ['Necromancer', 'm'],
    0x03 : ['Paladin', 'm'],
    0x04 : ['Barbarian', 'm'],
    0x05 : ['Druid', 'm'],
    0x06 : ['Assassin', 'f']
    }


diff = {
    'nor': {
        0x1: { 0 : { 'm': 'Sir', 'f': 'Dame' },
               1 : { 'm': 'Count', 'f': 'Countess' }
               },
               
        0x2: { 0 : { 'm': 'Lord', 'f': 'Lady' },
               1 : { 'm': 'Duke', 'f': 'Duchess' }
               },
               
        0x3: { 0 : { 'm': 'Baron', 'f': 'Baroness' },
               1 : { 'm': 'King', 'f': 'Queen' }
               }
        },

    'exp': {
        0x1: { 0 : { 'm': 'Slayer', 'f': 'Slayer' },
               1 : { 'm': 'Destroyer', 'f': 'Destroyer' }
               },
               
        0x2: { 0 : { 'm': 'Champion', 'f': 'Champion' },
               1 : { 'm': 'Conqueror', 'f': 'Conqueror' }
               },
               
        0x3: { 0 : { 'm': 'Patriarch', 'f': 'Matriarch' },
               1 : { 'm': 'Guardian', 'f': 'Guardian' }
               }
        }
    }




def remove_null(text):
    return text.split(chr(0).encode('utf-8'))[0].decode('utf-8')


def get_ladder(file):
    try:
        size = stat(file)[6]
        data = open(file, "rb")
    except:
        print("Error opening " + file + "for read")
        exit()
    
    maxtype, checksum = unpack(LD_HEAD, data.read(szLD_HEAD))

    size = size - szLD_HEAD

    head = []

    for i in range(maxtype):
        type, offset, number = unpack(LD_INDEX, data.read(szLD_INDEX))
        size = size - szLD_INDEX
        head.append(
        {
        'type': type,
        'offset': offset,
        'number': number
        })
        

    ladder = {}
    ladder['nor'] = []
    ladder['exp'] = []

    temp = {}
    temp['nor'] = []
    temp['exp'] = []


    while size > 0:
        try:
            experience, status, level, _class, charname = unpack(LD_INFO, data.read(szLD_INFO))
        except:
            ### Bad data
            size = size - szLD_INFO
            continue
        
        size = size - szLD_INFO

        ## Avoid null chars
        if not experience:
            continue
        
        charname = remove_null(charname)
        died = 0

        if status & S_EXP:
            _type = 'exp'
            difficulty = ((status >> 0x08) & 0x0f) / 5
        else:
            _type = 'nor'
            difficulty = ((status >> 0x08) & 0x0f) / 5

        if status & S_HC:
            hc = 1
            if status & S_DEAD:
                died = 1
        else:
            hc = 0
        
        c_class = classes[_class]

        if difficulty and diff[_type].has_key(difficulty):
            prefix = diff[_type][difficulty][hc][c_class[1]]
        else:
            prefix = None
        
        char = (experience, {
            'charname'   : charname,
            'prefix'     : prefix,
            'experience' : experience,
            'class'      : c_class[0],
            'sex'        : c_class[0],
            'level'      : level,
            'type'       : _type,
            'difficulty' : difficulty,
            'hc'         : hc,
            'died'       : died
            })
        ## Dupe char? why?
        if char not in temp[_type]:
            temp[_type].append(char)
        
    data.close()

    ## Sorting by exp
    temp['nor'].sort()
    temp['nor'].reverse()
    temp['exp'].sort()
    temp['exp'].reverse()

    for _type in temp.keys():
        for ch in temp[_type]:
            ladder[_type].append(ch[1])
    del temp

    return ladder

def parseLadder():
    file = "D:/PvPGN/Magic_Builder/release/var/ladders/ladder.D2DV"
    ladder = get_ladder(file)
    return ladder

File: Server API/LadderParser.go
package main

import (
	"encoding/binary"
	"encoding/json"
	"fmt"
	"os"
	"sort"
	"strings"
)

const LD_HEAD = "<2i"
const szLD_HEAD = 8 //calcsize(LD_HEAD)

const LD_INFO = "<Ihbb16s"
const szLD_INFO = 24 //calcsize(LD_INFO)

const LD_INDEX = "<3i"
const szLD_INDEX = 12 //calcsize(LD_INDEX)

const S_INIT = 0x1
const S_EXP = 0x20
const S_HC = 0x04
const S_DEAD = 0x08

var arr = [2]string{"Amazon", "f"}

var classes = map[int][]string{
	0x00: {"Amazon", "f"},
	0x01: {"Sorceress", "f"},
	0x02: {"Necromancer", "m"},
	0x03: {"Paladin", "m"},
	0x04: {"Barbarian", "m"},
	0x05: {"Druid", "m"},
	0x06: {"Assassin", "f"}}

var diff = map[string]map[int]map[int]map[string]string{
	"nor": {
		0x1: {0: {"m": "Sir", "f": "Dame"},
			1: {"m": "Count", "f": "Countess"}},

		0x2: {0: {"m": "Lord", "f": "Lady"},
			1: {"m": "Duke", "f": "Duchess"}},

		0x3: {0: {"m": "Baron", "f": "Baroness"},
			1: {"m": "King", "f": "Queen"}}},

	"exp": {
		0x1: {0: {"m": "Slayer", "f": "Slayer"},
			1: {"m": "Destroyer", "f": "Destroyer"}},

		0x2: {0: {"m": "Champion", "f": "Champion"},
			1: {"m": "Conqueror", "f": "Conqueror"}},

		0x3: {0: {"m": "Patriarch", "f": "Matriarch"},
			1: {"m": "Guardian", "f": "Guardian"}}}}

func remove_null(text string) string {
	return strings.Split(text, "\x00")[0]
}

/*
type character struct {
	charname   string
	prefix     string
	experience uint32
	class      string
	sex        string
	level      int
	_type      string
	difficulty int
	hc         int
	died       int
}
*/

type character struct {
	Charname   string `json:"charname"`
	Prefix     string `json:"prefix"`
	Experience uint32 `json:"experience"`
	Class      string `json:"class"`
	Sex        string `json:"sex"`
	Level      int    `json:"level"`
	Type       string `json:"type"`
	Difficulty int    `json:"difficulty"`
	Hc         int    `json:"hc"`
	Died       int    `json:"died"`
}

type ByExperience []character

func (a ByExperience) Len() int {
	return len(a)
}

func (a ByExperience) Less(i, j int) bool {
	return a[i].Experience > a[j].Experience
}

func (a ByExperience) Swap(i, j int) {
	a[i], a[j] = a[j], a[i]
}

func reverse(a []character) {
	for i, j := 0, len(a)-1; i < j; i, j = i+1, j-1 {
		a[i], a[j] = a[j], a[i]
	}
}

func get_ladder(file string) string {
	f, _ := os.Stat(file)
	size := f.Size()
	data, _ := os.Open(file)

	bs := make([]byte, 8)
	_, err := data.Read(bs)
	if err != nil {
		fmt.Println("ERROR!", err)
		return "Error"
	}
	maxtype := int(int32(binary.LittleEndian.Uint32(bs[:4])))
	//checksum := int(int32(binary.LittleEndian.Uint32(bs[4:])))

	size = size - szLD_HEAD

	var head []map[string]int

	for i := 0; i < maxtype; i++ {
		bs = make([]byte, 12)
		_, err := data.Read(bs)
		if err != nil {
			fmt.Println("ERROR!", err)
			return "ERROR"
		}
		_type := int(int32(binary.LittleEndian.Uint32(bs[:4])))
		offset := int(int32(binary.LittleEndian.Uint32(bs[4:8])))
		number := int(int32(binary.LittleEndian.Uint32(bs[8:])))
		size = size - szLD_INDEX
		head = append(head, map[string]int{"type": _type, "offset": offset, "number": number})
	}

	temp := make(map[string][]character)

	written := false
	for size > 0 {
		if len(temp["nor"]) > 0 && !written {
			fmt.Println("size: ", size)
			fmt.Println("\ttemp: ", temp["nor"])
			written = true
		}
		bs := make([]byte, 24)
		_, err := data.Read(bs)
		if err != nil {
			// Bad data
			size = size - szLD_INFO
			continue
		}

		experience := binary.LittleEndian.Uint32(bs[:4])
		status := int(int16(binary.LittleEndian.Uint16(bs[4:6])))
		level := int(bs[6])
		_class := int(bs[7])
		charname := string(bs[8:24])

		size = size - szLD_INFO

		// Avoid null chars
		if experience == 0 {
			continue
		}

		charname = remove_null(charname)
		died := 0
		_type := ""
		difficulty := 1
		hc := 0
		if status&S_EXP != 0 {
			_type = "exp"
			difficulty = ((status >> 0x08) & 0x0f) / 5
		} else {
			_type = "nor"
			difficulty = ((status >> 0x08) & 0x0f) / 5
		}
		if status&S_HC != 0 {
			hc = 1
			if status&S_DEAD != 0 {
				died = 1
			}
		} else {
			hc = 0
		}

		c_class := classes[_class]
		prefix := ""
		if _, ok := diff[_type][difficulty]; difficulty != 0 && ok == true {
			prefix = diff[_type][difficulty][hc][c_class[1]]
		} else {
			prefix = "a"
		}

		char := character{charname, prefix, experience, c_class[0], c_class[0], level, _type, difficulty, hc, died}

		// Dupe char? why?
		isExists := false
		for _, val := range temp[_type] {
			if val == char {
				isExists = true
				break
			}
		}
		if !isExists {
			temp[_type] = append(temp[_type], char)
		}
	}

	defer data.Close()

	sort.Sort(ByExperience(temp["nor"]))
	sort.Sort(ByExperience(temp["exp"]))

	ladder, err := json.Marshal(temp)
	if err != nil {
		fmt.Println("ERROR", err)
	}

	return string(ladder)
}

func parseLadder() string {
	file := "D:/PvPGN/Magic_Builder/release/var/ladders/ladder.D2DV"
	ladder := get_ladder(file)
	println(ladder)
	return ladder
}
